### PR TITLE
fix(callbox): check if there are no participants in call

### LINKED ISSUE #961: Incoming call is shown for old notifications
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

---

## Steps to reproduce
1. Have an old/missed call notification in your notifications
2. Have enough "new" notifications, that the call notification is not received on the `/notifications` endpoint
3. Make sure that the call notification gets back to the first 25 elements and therefore returned by the server (it's not happening all the time, not sure if we track the last notification id or the like)

### Expected behaviour
No call notification should be triggered by an old call notification

### Actual behaviour
Image

Image

## Diagnosis and logs

<details>
<summary>Diagnosis report</summary>

### Diagnosis report

| **Nextcloud Talk Desktop**      |                                                                                                                       |
| ------------------------------: | --------------------------------------------------------------------------------------------------------------------- |
| **Version**                     | v1.0.0                                                                                                                |
| **Built-in Talk version**       | v20.1.0                                                                                                               |
| **Operating system**            | Darwin 24.1.0 (Darwin Kernel Version 24.1.0: Thu Oct 10 21:03:15 PDT 2024; root:xnu-11215.41.3~2/RELEASE_ARM64_T6000) |
| **Executable Path**             | /Applications/Nextcloud Talk.app/Contents/MacOS/Nextcloud Talk                                                        |
| ****Connected to****            | -                                                                                                                     |
| **Nextcloud server version**    | 31.0.0 dev                                                                                                            |
| **Nextcloud Talk version**      | 21.0.0-dev.3                                                                                                          |
| **`notifications` app enabled** | ✅ yes                                                                                                                 |
| **`notify_push` app enabled**   | ✅ yes                                                                                                                 |

#### Application config

```json
{
  "theme": "default",
  "systemTitleBar": false,
  "monochromeTrayIcon": true,
  "zoomFactor": 1,
  "playSoundChat": "respect-dnd",
  "playSoundCall": "respect-dnd",
  "enableCallbox": "respect-dnd"
}
```

</details>

### PR DESCRIPTION
**Pull request by Antreesy**
### ☑️ Resolves

* Ref #961
* If there are no participants in call, call is already missed, and callbox shouldn't be shown
* If call is missed, normal notification with normal sound should be shown

### 🖼️ Screenshots

🏚️ Before | 🏡 After
---|---
Callbox all the time | "You missed a call" notification
Call ringtone | Normal beep

### 🚧 Tasks

- [ ] Test

### CODE REVIEW

File: src/callbox/renderer/callbox.service.ts [hunk 1]
```diff
@@ -49,9 +51,31 @@ async function hasCurrentUserJoinedCall(token: string) {
 		throw new Error('Cannot check whether current join the call - no current user found')
 	}
 	const participants = await getCallParticipants(token)
+	if (!participants.length) {
+		throw new Error('There are currently no participants in the call', { cause: MISSED_CALL })
+	}
```
Comment (DorraJaouad) [inline src/callbox/renderer/callbox.service.ts:56]: I am not if this should be treated as an error.

File: src/callbox/renderer/callbox.service.ts [hunk 2]
```diff
@@ -76,6 +100,10 @@ export function waitCurrentUserHasJoinedCall(token: string, limit?: number): Pro
 					return resolve(true)
 				}
 			} catch (e) {
+				if (e instanceof Error && e.cause === MISSED_CALL) {
```
Comment (DorraJaouad) [inline src/callbox/renderer/callbox.service.ts:103]: if `isCallstillActive` returns falsy, `return resolve(false)`

File: src/callbox/renderer/callbox.service.ts (hunk 1 above)
Comment (Antreesy) [reply src/callbox/renderer/callbox.service.ts:56]: Either that or we change original method in first PR to handle 4 cases:
1) Call pending, not joined - status 200 - loop again
2) Call ended, not joined - status 201 - resolve(false) - dismiss and show 'missed call'
3) Call pending, already joined - status 404 - dismiss and show nothing
4) Other error - ?

File: src/callbox/renderer/callbox.service.ts [hunk 3]
```diff
@@ -49,9 +51,31 @@ async function hasCurrentUserJoinedCall(token: string) {
...
+	}
 	return participants.some((participant) => user.uid === participant.actorId)
 }
 
+/**
+ * Check if callbox should be rendered
+ * @param token - Conversation token
+ * @return Promise<boolean> - Resolved with boolean - true if the user should see the callbox, false otherwise
+ */
+export async function checkCurrentUserHasPendingCall(token: string): Promise<boolean> {
```
[the remark this answers is not in the document]
Comment (nickvergessen) [reply src/callbox/renderer/callbox.service.ts:65]: This is all quite temporary. The performance impact on the server side is HUGE.
We can not run more than 1 request every X seconds. `getPeers` is the approach that should currently be taken and in the future https://github.com/nextcloud/spreed/pull/13930 should be used.

File: src/callbox/renderer/callbox.service.ts (hunk 1 above)
Comment (Antreesy) [reply src/callbox/renderer/callbox.service.ts]: Replaced with `null`, should be enough for current implementation

**Review by DorraJaouad — APPROVED** [review]
Tested

**Pull request merged**